service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/CreateAzureVmStep.java

```diff
@@ -0,0 +1,250 @@
+        logger.info(
+            "Either the disk, ip, or network passed into this createVm does not exist "
+                + String.format(
+                    "\nResource Group: %s\n\tIp Name: %s\n\tNetwork Name: %s\n\tDisk Name: %s",
+                    azureCloudContext.getAzureResourceGroupId(),
+                    ipResource.getIpName(),
+                    "TODO",
+                    diskResource.getDiskName()));
+      }
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_FATAL, e);
+    }
+
+    return StepResult.getStepResultSuccess();
+  }
+
+  @Override
+  public StepResult undoStep(FlightContext context) throws InterruptedException {
+    ComputeManager computeManager = crlService.getComputeManager(azureCloudContext, azureConfig);
+
+    try {
+      computeManager
+          .virtualMachines()
+          .deleteByResourceGroup(azureCloudContext.getAzureResourceGroupId(), resource.getVmName());
+    } catch (ManagementException e) {
+      // Stairway steps may run multiple times, so we may already have deleted this resource.
+      if (StringUtils.equals(e.getValue().getCode(), "ResourceNotFound")) {
+        logger.info(
+            "Azure VM {} in managed resource group {} already deleted",
+            resource.getVmName(),
+            azureCloudContext.getAzureResourceGroupId());
+        return StepResult.getStepResultSuccess();
+      }
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, e);
+    }
+    return StepResult.getStepResultSuccess();
+  }
+
+  // TODO: THIS IS NOT HOW NETWORKS WILL BE CREATED, AND WILL BE REMOVED TO USE A CONTROLLED
+  // RESOURCE ONCE THAT WORK IS DONE (SEPARATE TICKET)
+  // Most of this is taken from this very helpful snippet
+  // https://github.com/Azure-Samples/network-java-manage-virtual-network/blob/master/src/main/java/com/azure/resourcemanager/network/samples/ManageVirtualNetwork.java
+  private Network createNetwork(
+      String networkName,
+      String subnetName,
+      String addressSpaceCidr,
+      String subnetAddressSpaceCidr) {
+    ComputeManager computeManager = crlService.getComputeManager(azureCloudContext, azureConfig);
+
+    NetworkSecurityGroup subnetNsg =
+        computeManager
+            .networkManager()
+            .networkSecurityGroups()
+            .define(subnetName)
+            .withRegion(resource.getRegion())
+            .withExistingResourceGroup(azureCloudContext.getAzureResourceGroupId())
+            .withTag("workspaceId", resource.getWorkspaceId().toString())
+            .withTag("resourceId", resource.getResourceId().toString())
+            .defineRule("AllowHttpInComing")
+            .allowInbound()
+            .fromAddress("INTERNET")
+            .fromAnyPort()
+            .toAnyAddress()
+            .toPort(8080)
+            .withProtocol(SecurityRuleProtocol.TCP)
+            .attach()
+            .defineRule("DenyInternetOutGoing")
```

Comment:
I think we should remove this `DenyInternetOutGoing` rule.. we don't want to restrict outgoing traffic.

Reply:
can we remove this